src/parallel-rustc.md

```diff
@@ -1,34 +1,117 @@
+
+| data structure                   | parallel                                            | non-parallel |
+| -------------------------------- | --------------------------------------------------- | ------------ |
+| Lrc                              | std::sync::Arc | std::rc::Rc |
+| Weak                             | std::sync::Weak                                     | std::rc::Weak |
+| Atomic{Bool}/{Usize}/{U32}/{U64} | std::sync::atomic::Atomic{Bool}/{Usize}/{U32}/{U64} | (std::cell::Cell<bool/usize/u32/u64>) |
+| OnceCell                         | std::sync::OnceLock                                 | std::cell::OnceCell |
+| Lock\<T> | (parking_lot::Mutex\<T>) | (std::cell::RefCell) |
+| RwLock\<T> | (parking_lot::RwLock\<T>) | (std::cell::RefCell) |
+| MTRef<'a, T> | &'a T | &'a mut T |
+| MTLock\<T> | (Lock\<T>) | (T) |
+| ReadGuard | parking_lot::RwLockReadGuard | std::cell::Ref |
+| MappedReadGuard | parking_lot::MappedRwLockReadGuard | std::cell::Ref |
+| WriteGuard | parking_lot::RwLockWriteGuard | std::cell::RefMut |
+| MappedWriteGuard | parking_lot::MappedRwLockWriteGuard | std::cell::RefMut |
+| LockGuard | parking_lot::MutexGuard | std::cell::RefMut |
+| MappedLockGuard | parking_lot::MappedMutexGuard | std::cell::RefMut |
+| MetadataRef | [`OwningRef<Box<dyn Erased + Send + Sync>, [u8]>`][OwningRef] | [`OwningRef<Box<dyn Erased>, [u8]>`][OwningRef] |
+
+- These thread-safe data structures interspersed during compilation can 
+  cause a lot of lock contention, which actually degrades performance as the
+  number of threads increases beyond 4. This inspires us to audit the use 
+  of these data structures, leading to either refactoring to reduce use of 
+  shared state, or persistent documentation covering invariants, atomicity, 
+  and lock orderings.
+
+- On the other hand, we still need to figure out what other invariants 
+  during compilation might not hold in parallel compilation.
+
+### WorkLocal
+
+`WorkLocal` is a special data structure implemented for parallel compiler.
+It holds worker-locals values for each thread in a thread pool. You can only 
+access the worker local value through the Deref impl on the thread pool it 
+was constructed on. It will panic otherwise. 
+
+`WorkLocal` is used to implement the `Arena` allocator in the parallel 
+environment, which is critical in parallel queries. Its implementation 
+is located in the `rustc-rayon-core::worker_local` module. However, in the 
+non-parallel compiler, it is implemented as `(OneThread<T>)`, whose `T` 
+can be accessed directly through `Deref::deref`.
+
+## Parallel Iterator
+
+The parallel iterators provided by the [`rayon`] crate are easy ways 
+to implement parallelism. In the current implementation of the parallel 
+compiler we use a custom fork of [`rayon`] to run tasks in parallel. 
+*(more information wanted here)*
+
+Some iterator functions are implemented in the current nightly compiler to
```

Comment:
```suggestion
Some iterator functions are implemented to
```
The default implication is all docs refer to current compiler, unless specified.
Also, the *nightly* can be dropped for reason stated at https://github.com/rust-lang/rustc-dev-guide/pull/1432#discussion_r951963151.